### PR TITLE
Add ARM Compiler 6 build tests to all.sh

### PR DESCRIPTION
**Pull request by andresag01**
I would like to point out that the script `tests/script/yotta-build.sh` actually uses armcc. However, I do not think there are yotta targets for ARM Compiler 6. Furthermore, we might need to evaluate how to force yotta to use a specific arm compiler binary.

### CODE REVIEW

File: scripts/output_env.sh [hunk 1]
```diff
@@ -22,11 +22,30 @@ echo "* Operating system and architecture:"
...
+    fi
+fi
+
+if [ -n "${ARMC6_CC+set}" ]; then
+    echo
+    if `hash "$ARMC6_CC" > /dev/null 2>&1`; then
+        echo "* $ARMC6_CC at environment variable 'ARMC6_CC':"
+        $ARMC6_CC --vsn | head -n 2
+    else
+        echo "* $ARMC6_CC at environment variable 'ARMC6_CC' not found!"
```
Comment (sbutcher-arm) [inline scripts/output_env.sh]: Above you have `hash armcc`, so why don't you test the presence of ARM C Compiler v6 with `hash armclang`?

The point of the environment variables is to specify the full path to the tool, but `armcc` and `armclang` may both be already on the path.

File: tests/scripts/all.sh [hunk 2]
```diff
@@ -41,6 +41,8 @@ FORCE=0
 : ${GNUTLS_LEGACY_CLI:="$GNUTLS_CLI"}
 : ${GNUTLS_LEGACY_SERV:="$GNUTLS_SERV"}
 : ${OUT_OF_SOURCE_DIR:=./mbedtls_out_of_source_build}
+: ${ARMC5_BIN_DIR:=/usr/bin}
+: ${ARMC6_BIN_DIR:=/usr/bin}
```
Comment (sbutcher-arm) [inline tests/scripts/all.sh]: You're looking for the path to the compilers, and defaulting to /usr/bin.

I would suggest you find the full path of `armcc` and `armclang` using `which`, then extract the directory using `dirname`. (Or do something equivalent).

File: tests/scripts/all.sh [hunk 3]
```diff
@@ -55,6 +57,8 @@ usage()
     printf "     --gnutls-serv=<GnuTLS_serv_path>\t\tPath to GnuTLS server executable to use for most tests.\n"
     printf "     --gnutls-legacy-cli=<GnuTLS_cli_path>\t\tPath to GnuTLS client executable to use for legacy tests.\n"
     printf "     --gnutls-legacy-serv=<GnuTLS_serv_path>\t\tPath to GnuTLS server executable to use for legacy tests.\n"
+    printf "     --armc5-bin-dir=<ARMC5_bin_dir_path>\t\tPath to the ARM Compiler 5 bin directory.\n"
```
Comment (sbutcher-arm) [inline tests/scripts/all.sh]: I don't think we should take the directory, but similar to OpenSSL and GNUTLS for consistency, the full path for `armcc` and `armclang`,

File: tests/scripts/all.sh [hunk 4]
```diff
@@ -403,7 +423,19 @@ scripts/config.pl unset MBEDTLS_THREADING_C
 scripts/config.pl unset MBEDTLS_MEMORY_BACKTRACE # execinfo.h
 scripts/config.pl unset MBEDTLS_MEMORY_BUFFER_ALLOC_C # calls exit
 scripts/config.pl unset MBEDTLS_PLATFORM_TIME_ALT # depends on MBEDTLS_HAVE_TIME
-CC=armcc AR=armar WARNING_CFLAGS= make lib
+
+CC="$ARMC5_CC" AR="$ARMC5_AR" WARNING_CFLAGS= make lib
```
Comment (sbutcher-arm) [inline tests/scripts/all.sh]: I think the CFLAGS below should be tighter. Fail on warnings, pedantic, strict etc.

File: scripts/output_env.sh (hunk 1 above)
Comment (andresag01) [inline scripts/output_env.sh]: I can do `hash armclang` if your prefer. However, I do not think it is sensible to have both `armcc` and `armclang` in the path simultaneously. This is because these toolchains have executables with names that overlap. For instance, `armar` and `armlink` are called the same in both cases. If you have both bin folders in the path, then you will potentially be compiling with `armclang` or `armcc` as desired BUT you might be using the wrong version of other tools such as `armar`. I tried this and it definitely fails.

File: tests/scripts/all.sh (hunk 4 above)
Comment (andresag01) [inline tests/scripts/all.sh]: I agree with this comment. However, I would also like to point out that there is a separate PR https://github.com/ARMmbed/mbedtls/pull/585 that already does this to some extent. I consider that we should do this change as a separate PR or add to the existing PR https://github.com/ARMmbed/mbedtls/pull/585 rather than introduce the changes here.

**Comment by yanesca** [comment]
Looks good to me.

**Comment by sbutcher-arm** [comment]
Reviewed and approved. Needs testing then merge.

**Comment by andresag01** [comment]
@sbutcher-arm: I am labeling this as `needs work` once more. I tested the patch and the compiler throws the following warning:

```
warning: warning: warning: warning: warning: warning: 'armv4t' is unsupported in this version of the product'armv4t' is unsupported in this version of the product
warning: 'armv4t' is unsupported in this version of the product
'armv4t' is unsupported in this version of the productwarning: 'armv4t' is unsupported in this version of the product
```

I need to check the compiler documentation to work out what the problem is.

**Comment by andresag01** [comment]
Fixed the warning by passing the `-march` flag to `armclang`. Removing DRAFT and `needs rework` labels.

File: tests/scripts/all.sh (hunk 2 above)
Comment (sbutcher-arm) [inline tests/scripts/all.sh]: Ok. Agreed.

File: tests/scripts/all.sh (hunk 3 above)
Comment (sbutcher-arm) [inline tests/scripts/all.sh]: I now accept it as it is.

File: tests/scripts/all.sh (hunk 4 above)
Comment (sbutcher-arm) [inline tests/scripts/all.sh]: Point accepted. We now have #695 to address this.

**Comment by sbutcher-arm** [comment]
LGTM. Needs testing before merge.

**Comment by sbutcher-arm** [comment]
I missed this in review, but in testing, I found the output from the script where we check what tools are present isn't as clear as before - and for the OpenSSL/GnuTLS commands, the legacy tools aren't obvious enough.

Whereas before we output something like:
```
 * armcc: Product: ARM Compiler 5.06 Component: ARM Compiler 5.06 update 4 (build 422)
```
We now output the full path of the tool:
```
* /usr/local/ARM_Compiler_5.06u4/bin//armcc: Product: ARM Compiler 5.06 Component: ARM Compiler 5.06 update 4 (build 422)
```
Can we please prefix each of the results with a 'simple name' of the tool, in addition to the command that will be executed? And can that 'simple name' capture the purpose, if it's a specific legacy version?

eg.
```
* armcc 5: /usr/local/ARM_Compiler_5.06u4/bin//armcc: Product: ARM Compiler 5.06 Component: ARM Compiler 5.06 update 4 (build 422)
```

**Comment by andresag01** [comment]
@sbutcher-arm: I agree that printing the full path to the binary first might make the output harder to read. Therefore, I have modified the output slightly to output in the format:

```
<PROGRAM_NAME>: <BINARY_PATH>: <VERSION_OUTPUT>
```
Where:

* `<PROGRAM_NAME>` is the basename of the binary path. For example, in `foo/bar/vaz` the program name would be `vaz`.
* `<BINARY_PATH>` is the full path to the program binary as returned by the `which` shell utility.
* `<VERSION_OUTPUT>` is the actual version output given by the program. For example, in the case of GCC it would be the output for `gcc --version`.

Finally, you mentioned that the `<PROGRAM_NAME>` should `capture the purpose, if it's a specific legacy version?`. In the case of `armcc` you would like to print `armcc 5`, but what should we print in the case of other applications like gnutls, openssl, clang, valgrind, etc. Note that getting the exact legacy version is what we are trying to do as part of the `<VERSION_OUTPUT>` anyways and extracting this information may not be straight forward since output of the applications do not follow a specific format.

**Comment by sbutcher-arm** [comment]
Reviewed, tested, merged.
